zerver/actions/message_flags.py

```diff
@@ -315,8 +317,18 @@ def do_update_message_flags(
             )
 
         if operation == "add":
+            messages = [
+                message_id
+                for message_id in messages
+                if ~um_flags.get(message_id, DEFAULT_HISTORICAL_FLAGS) & flagattr != 0
+            ]
```

Comment:
In an ideal world we'd have a comment here explaining that we're filtering for rows that already have the correct flags here, rather than in the original database query, because not all flags have database indexes and we want to bound the cost of this operation.

Also it's kinda a code smell that we have this line but then we use `query` for the UPDATE operation, which was constructed with the original set of messages. I'm not sure if it'd be helpful to do this `messages` update after the `UPDATE` query to make the intentionality more clear, or we should add a comment explaining that because this is in a transaction where we did `SELECT FOR UPDATE`, something something....

(I think we need to reuse the `query` variable given it's the thing we did a `select_for_update` on?)